### PR TITLE
Fix bugs and refactor codes

### PR DESCRIPTION
**Pull request by hy2850**
- pass copied tensor instead of reference
- refactored 'attention extension' code to more readable form
- fix bugs on 'scale_shared_grads'

### CODE REVIEW

File: utils/critics.py [hunk 1]
```diff
@@ -208,44 +215,66 @@ def forward(self, inps, agents=None, return_q=True, return_all_q=False,
...
-
-        # extend the cluster attentions to agent attentions
-        for clst_idx, c_agents in self.cluster_list.items():
+                # Bug!) pushing the same reference into the list
+                # -> all tensors are actually the same single tensor, so updates are shared
+                # clst_logits_extended[i].append(temp_logits)
+                # clst_probs_extended[i].append(temp_probs)
+                
+                # must copy and append
+                clst_logits_extended[i].append(temp_logits.detach().clone())
```
Comment (hy2850) [inline utils/critics.py:225]: 버그 수정 : pass copy, not reference

File: utils/critics.py [hunk 2]
```diff
@@ -208,44 +215,66 @@ def forward(self, inps, agents=None, return_q=True, return_all_q=False,
...
+                            # > For agents that appear after me, exclude myself from attention vector index value by subtracting 1
+                            if current_agent < other_agent:
+                                other_agent -= 1
+
+                            # Q. just assign? not add? -> each cell is accessed and copied only once
+                            # clst_logits_extended[a_c][i_head][:, :, c_agent] = clst_attend_logits[clst_idx][i_head][:, :, other_clst]
+                            # clst_probs_extended[a_c][i_head][:, :, c_agent] = clst_attend_probs[clst_idx][i_head][:, :, other_clst]
+                            clst_probs_extended[current_agent][i_head][:, :, other_agent] = clst_attend_probs[current_clst][i_head][:, :, other_clst]
+                            clst_logits_extended[current_agent][i_head][:, :, other_agent] = clst_attend_logits[current_clst][i_head][:, :, other_clst]
+                            # Why not 1? Shouldn't clst_probs_extended[0][0][0][0][0~4] be changed and stay as 1?
```
Comment (hy2850) [inline utils/critics.py:260]: 유승님이 의도하신 바를 유지하면서 읽기 쉽게 리팩토링

File: utils/cluster_critics.py [hunk 3]
```diff
@@ -97,7 +97,10 @@ def scale_shared_grads(self):
         gradients from the critic loss function multiple times
         """
         for p in self.shared_parameters():
-            p.grad.data.mul_(1. / self.n_clusters)
+            if p.grad == None:
```
Comment (hy2850) [inline utils/cluster_critics.py:100]: grad None이면 일반 data라도 normalize
(self.shared_modules 중, self.c_value_extractors에서 이 문제가 발생하는데, 이유 못찾음)

**Comment by hy2850** [comment]
cluster creation code has been merged onto this branch; code runs well
